compiler/rustc_serialize/src/opaque.rs

```diff
@@ -511,34 +512,89 @@ impl Encoder for FileEncoder {
 // -----------------------------------------------------------------------------
 
 pub struct MemDecoder<'a> {
-    pub data: &'a [u8],
-    position: usize,
+    start: *const u8,
+    current: *const u8,
+    end: *const u8,
+    _marker: PhantomData<&'a u8>,
 }
 
 impl<'a> MemDecoder<'a> {
     #[inline]
     pub fn new(data: &'a [u8], position: usize) -> MemDecoder<'a> {
-        MemDecoder { data, position }
+        let range = data.as_ptr_range();
+        MemDecoder {
+            start: range.start,
+            current: data[position..].as_ptr(),
+            end: range.end,
+            _marker: PhantomData,
+        }
+    }
+
+    #[inline]
+    pub fn data(&self) -> &'a [u8] {
+        unsafe { std::slice::from_raw_parts(self.start, self.len()) }
+    }
+
+    #[inline]
+    pub fn len(&self) -> usize {
+        unsafe { self.end.sub_ptr(self.start) as usize }
```

Comment:
nit: <https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.sub_ptr> returns a `usize` -- that's a huge part of its reason to exist :)

```suggestion
        unsafe { self.end.sub_ptr(self.start) }
```